opm/material/fluidsystems/BlackOilFluidSystem.hpp

```diff
@@ -313,12 +313,12 @@ class BlackOil : public BaseFluidSystem<Scalar, BlackOil<Scalar> >
         switch (phaseIdx) {
         case waterPhaseIdx: return waterDensity<LhsEval>(T, p, regionIdx);
         case gasPhaseIdx: {
-            const auto& XgO = FsToolbox::template toLhs<LhsEval>(fluidState.massFraction(gasPhaseIdx, oilCompIdx));
-            return gasDensity<LhsEval>(T, p, XgO, regionIdx);
+            const auto& Rv = getRv_<LhsEval>(fluidState, regionIdx);
+            return gasDensity<LhsEval>(T, p, Rv, regionIdx);
         }
         case oilPhaseIdx: {
-            const auto& XoG = FsToolbox::template toLhs<LhsEval>(fluidState.massFraction(oilPhaseIdx, gasCompIdx));
-            return oilDensity<LhsEval>(T, p, XoG, regionIdx);
+            const auto& Rs = getRs_<LhsEval>(fluidState, regionIdx);
```

Comment:
It seems to me that you intend for the fluid state class to still use mass fraction, even while changing the rest of this system to use Rs/Rv. Is that not a bit unfortunate? Is it for backwards compatibility, or compatibility with other (not blackoil) models? I guess it is hard (but perhaps not impossible?) to make a code that works with Rs if available and mass fraction if not.